udffsck/udffsck.c

```diff
@@ -0,0 +1,3370 @@
+        prot |= PROT_WRITE;
+        dbg("\tRW\n");
+    }
+
+    *ptr = (uint8_t *)mmap(NULL, size, prot, MAP_SHARED, fd, offset);
+    if(ptr == MAP_FAILED) {
+        switch(errno) {
+            case EACCES: dbg("EACCES\n"); break;
+            case EAGAIN: dbg("EAGAIN\n"); break;
+            case EBADF: dbg("EBADF\n"); break;
+            case EINVAL: dbg("EINVAL\n"); break;
+            case ENFILE: dbg("ENFILE\n"); break;
+            case ENODEV: dbg("ENODEV\n"); break;
+            case ENOMEM: dbg("ENOMEM\n"); break;
+            case EPERM: dbg("EPERM\n"); break;
+            case ETXTBSY: dbg("ETXTBSY\n"); break;
+            case EOVERFLOW: dbg("EOVERFLOW\n"); break;
+            default: dbg("EUnknown\n"); break;
+        }
+
+        fatal("\tError maping: %s.\n", strerror(errno));
+        exit(8); 
+    }
+#ifdef MEMTRACE
+    dbg("\tArea allocated, pointer: %p, offset 0x%x\n", ptr, offset);
+#else
+    dbg("\tArea allocated\n");
+#endif
+}
+
+char * dstring_suberror(uint8_t e_code) {
+   switch(e_code) {
+        case 0:
+            return NULL;
+        case DSTRING_E_NONZERO_PADDING:
+            return "non-zero padding";
+        case DSTRING_E_WRONG_LENGTH:
+            return "wrong length";
+        case DSTRING_E_INVALID_CHARACTERS:
+            return "invalid characters present";
+        case DSTRING_E_NOT_EMPTY:
+            return "string is not empty";
+        case DSTRING_E_UNKNOWN_COMP_ID:
+            return "unknown Compression ID";
+        default: 
+            return "unknown dstring error";
+   } 
+}
+
+uint8_t dstring_error(char * string_name, uint8_t e_code) {
+    char * buffer[600];
+    if(e_code > 0) {
+        msg("Dstring %s has following errors:\n", string_name);
+        for(int i=0; i<8; ++i) {
+            if(e_code & 1<<i) {
+                msg("\t- %s\n", dstring_suberror(e_code & 1<<i));
+            }
+        }
+
+        return 4; 
+    }
+    return 0;
+}
+
+/**
+ * \brief Function for detection of errors in dstrings
+ *
+ * This function detects violation against UDF 2.1.1 in dstrings
+ * Theese violations are:
+ *  1. Non-zero padding
+ *  2. Not allowed characters for 16 bit dchars (0xFFFE and 0xFEFF)
+ *  3. Unknown compression ID
+ *  4. Emptyness of string if ID is 0 or length is 0
+ *  5. String legth mismatch
+ *
+ * \param[in] in dstring for check
+ * \param[in] field_size size of dstring field
+ * \return sum of DSTRING_E codes 
+ */
+uint8_t check_dstring(dstring *in, size_t field_size) {
+    uint8_t compID = in[0];
+    uint8_t length = in[field_size-1];
+    uint8_t stepping = 0xFF;
+    uint8_t empty_flag = 0;
+    uint8_t e_code = 0;
+
+    dbg("compID: %d, length: %d\n", compID, length);
+    switch(compID) {
+        case 8: 
+            stepping = 1;
+            break;
+        case 16: 
+            stepping = 2;
+            break;
+        case 0:
+            stepping = 1;
+            empty_flag = 1;
+            break;
+        default:
```

Comment:
There are another two specials compression ids: 254 and 255.

Reply:
Oh, I see. I forgot to check 2.01 standard. 

Although, I am not sure how to interpret it.

> **UDF 2.01, 2.1.1**
Compression IDs 254 and 255 shall only be used in FIDs where the deleted bit is set to
ONE.
When uncompressing file identifiers with Compression IDs 254 and 255, the resulting
name is to be considered empty and unique. 

But FID has no dstring defined for name. There is only string (char array) and it is _FileIdentifier_. It has also separate field _LengthofFileIdentifier_. I am not sure how to handle this. Also search in UDF 2.01 spec didn't shown any other dstrings than those implemented.

Reply:
Right, there are only `d-characters` where is compression id 254 and 255 used. So for validation of `d-characters` is needed slightly modified version as for `d-string`.

Reply:
I am sorry, but I still do not follow you on this. 

Shall I interpret char array at FID as dstring or not?

Reply:
There are two types: d-string and d-characters and they differ just in last byte (d-string has stored there length). And so FID needs to be interpreted as d-characters.

Reply:
Oh, ok. I'll implement that.

Reply:
Ok, If all test will be green, implementation is here:  96a0d4c 
There is slight possibility that deleted FIDs will have some errors inside, so I am looking forward for results of tests.

Btw, during implementation of dchars I found interesting thing - only detectable error for them except from compression ID are invalid characters for u16. Nothing else.